src/features/bivariate_manager/atoms/bivariateMatrixSelection.ts

```diff
@@ -38,10 +65,172 @@ export const bivariateMatrixSelectionAtom = createAtom(
       xDenominator: string | null;
       yNumerator: string | null;
       yDenominator: string | null;
-    } = UNSELECTED_STATE,
+      preselectMode: boolean | false;
+      selectedCell: { x: number; y: number } | null;
+      selectCellCallback: SelectCellCallback | null;
+    } = DEFAULT_STATE,
   ) => {
+    onAction('calculateSelectedCell', () => {
+      const { xGroups, yGroups } = getUnlistedState(bivariateNumeratorsAtom);
+      const nextSelectedCell = onCalculateSelectedCell(
+        xGroups,
+        yGroups,
+        formatSelection(
+          state.xNumerator,
+          state.xDenominator,
+          state.yNumerator,
+          state.yDenominator,
+        ),
+      );
+      const currentCell = state.selectedCell;
+
+      if (
+        !currentCell ||
+        nextSelectedCell.x !== currentCell.x ||
+        nextSelectedCell.y !== currentCell.y
+      ) {
+        state = { ...state, selectedCell: nextSelectedCell };
+      }
+    });
+
+    // in these 2 onChange callbacks we react on selection of overlays
+    // enabledLayersAtom we need when we select layer already selected before - it has legend downloaded
+    onChange('enabledLayersAtom', () => {
+      if (state.preselectMode) {
+        schedule((dispatch) => {
+          dispatch(create('runPreselection'));
+        });
+      }
+    });
+    // layersLegendsAtom we need when we select layer first time, it has no legend, so we need to wait until it's ready
+    onChange('layersLegendsAtom', () => {
+      if (state.preselectMode) {
+        schedule((dispatch) => {
+          dispatch(create('runPreselection'));
+        });
+      }
+    });
+    // when matrix component mounted we set this mode and need to run selection first time
+    onAction('setPreselectMode', (preselectMode) => {
+      state = { ...state, preselectMode };
+      if (preselectMode) {
+        schedule((dispatch) => {
+          dispatch(create('runPreselection'));
+        });
+      }
+    });
+
+    // this is onSelect method from BivariateMatrixControl
+    onAction('setSelectCellCallback', (selectCellCallback) => {
+      state = { ...state, selectCellCallback };
+    });
+
+    const getActiveBivariateLayer = () => {
```

Comment:
`getMounterBivariateLayer` or rather `getEnabledBivariateLayer` would be more consistant